File: src/main/java/javaclasses/Algorithm/KadaneAlgorithm.java
/*
 * *
 *  * Created by Rafsan Ahmad on 10/25/21, 9:49 PM
 *  * Copyright (c) 2021 . All rights reserved.
 *
 */

package javaclasses.Algorithm;

public class KadaneAlgorithm {
    /*A simple idea of Kadane’s algorithm is to look for all positive contiguous segments of the array and keep
    track of the maximum sum contiguous subarray among all positive segments.

First, we will consider two elements, one which stores the maximum end of the subarray and another which stores
the maximum sum so far.

Let these two variables be max_ending_here and max_so_far, respectively.
We will initialise both of them to 0.
Each time we get a positive sum, we compare it with max_so_far and update max_so_far if it is greater than it.
This logic is written in the form of an algorithm as follows:

1. Start
2. max_so_far = 0
3. max_ending_here = 0
4. Loop for each element of the array
       1. max_ending_here = max_ending_here + a[i]
       2. if(max_ending_here < 0)
          max_ending_here = 0

       4. if(max_so_far < max_ending_here)
          max_so_far = max_ending_here

5. return max_so_far*/


    public int kadane(int[] arr, int n) {
        //stores maximum sum subarray found so far

        int max_so_far = 0;

        //stores the maximum sum of subarray ending at the current position

        int max_ending_here = 0;

        //traverse the given array

        for (int i = 0; i < n; i++) {
            /* update maximum sum of subarray "ending" at index i (by adding the current element to maximum sum ending at previous index i-1) */

            max_ending_here = max_ending_here + arr[i];

            /* if maximum sum is negative, set it to 0 (which represents an empty sub-array) */
            if (max_ending_here < 0) {
                max_ending_here = 0;
            }

            //update result if current subarray sum is found to be greater
            if (max_so_far < max_ending_here) {
                max_so_far = max_ending_here;
            }
        }
        return max_so_far;
    }

    public static void main(String[] args) {
        int[] arr = {-2, -3, 4, -1, -2, 1, 5, -3};
        int n = arr.length;
        System.out.println(new KadaneAlgorithm().kadane(arr, n));
    }


}

File: src/main/java/kotlinclasses/Algorithm/KadaneAlgorithm.kt
/*
 * *
 *  * Kadane Algorithm.kt
 *  * Created by Rafsan Ahmad on 6/30/25, 7:38PM
 *  * Copyright (c) 2025 . All rights reserved.
 *  *
 *
 */

package kotlinclasses.Algorithm

class KadaneAlgorithm {
    /*Kadane's Algorithm is a dynamic programming approach used to solve the problem:

Find the maximum sum of a contiguous subarray in a 1D array of integers.
It runs in O(n) time and O(1) space.

✅ Intuition
At each position in the array, you have two choices:
Start a new subarray from the current element.
Extend the previous subarray by including the current element.
You pick whichever gives you a larger sum.

✅ Step-by-Step Logic
Variables used:
currentSum: the maximum sum ending at the current index
maxSum: the overall maximum sum found so far

Steps:
Initialize:
currentSum = nums[0]
maxSum = nums[0]

Loop through the array starting from index 1 to the end:
currentSum = max(nums[i], currentSum + nums[i])

Choose the larger one: start new from nums[i] or add to previous sum
maxSum = max(maxSum, currentSum)

Update the overall max if needed
Return maxSum

    Example 1:
    Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
    Output: 6
    Explanation: [4,-1,2,1] has the largest sum = 6.

    Example 2:
    Input: nums = [1]
    Output: 1

    Example 3:
    Input: nums = [5,4,-1,7,8]
    Output: 23
*/

    //Time complexity: O(n)
    fun maxSubArray(nums: List<Int>): Int {
        if (nums.isEmpty()) return 0
        var currentSum = 0
        var maxSum = Integer.MIN_VALUE
        for (i in 0 until nums.size) {
            currentSum += nums[i]
            if (nums[i] > currentSum)
                currentSum = nums[i]
            maxSum = maxOf(maxSum, currentSum)
        }
        return maxSum
    }
}

fun main() {
    val obj = KadaneAlgorithm()
    println(obj.maxSubArray(listOf(-2, 1, -3, 4, -1, 2, 1, -5, 4)))
    println(obj.maxSubArray(listOf(5, -3, 5)))
    println(obj.maxSubArray(listOf(5, 4, -1, 7, 8)))
}